### ISSUE TITLE
exception object type

### ISSUE THREAD

**Issue by YexuanXiao** [YexuanXiao/basic_json#14]
Adjust the type of the exception object to class json_error instead of runtime_error. A possible implementation is as follows:

```cpp
class json_error: public runtime_error
{
public:
    enum class errc: unsigned int
    {
        /**/
    };

    errc code() const noexcept
    {
        return code_;
    }

    json_error(errc code) noexcept : code_(code)
    {
    }

    ~json_error() override
    {
    }

    json_error& operator=(const json_error& e) noexcept = default;

    const char* what() const noexcept override
    {
        /**/
        return "";
    }

private:
    errc code_{};
};
```

**Comment by frederick-vs-ja** [comment]
Should `errc` be nested in `json_error` or at the namespace scope?

**Comment by YexuanXiao** [comment]
> Should `errc` be nested in `json_error` or at the namespace scope?

I am considering using the latter. The former doesn’t seem to have any particular advantages and may be controversial.

**Comment by frederick-vs-ja** [comment]
I'll change the enumeration to `json_errc` at the namespace scoe.

I have to drop the `noexcept` from the `json_errc` ctor since the standard doesn't provide a `noexcept` way to construct the `runtime_error` object other than copying from an existing `runtime_error`.

**Comment by YexuanXiao** [comment]
> I'll change the enumeration to `json_errc` at the namespace scoe.
> 
> I have to drop the `noexcept` from the `json_errc` ctor since the standard doesn't provide a `noexcept` way to construct the `runtime_error` object other than copying from an existing `runtime_error`.

Unused error codes should be removed from the proposal at this time. Also, I'd recommend making the underlying type at least int, to leave plenty of room for possible extensions.

**Comment by frederick-vs-ja** [comment]
> > I'll change the enumeration to `json_errc` at the namespace scoe.
> > I have to drop the `noexcept` from the `json_errc` ctor since the standard doesn't provide a `noexcept` way to construct the `runtime_error` object other than copying from an existing `runtime_error`.
> 
> Unused error codes should be removed from the proposal at this time. Also, I'd recommend making the underlying type at least int, to leave plenty of room for possible extensions.

Addressed in 149c3026dbadb549522e257882713945d29e229e.
